src/binary/vector.rs

```diff
@@ -0,0 +1,288 @@
+use serde::{Deserialize, Serialize};
+
+use super::{Binary, Error, Result};
+use crate::{spec::BinarySubtype, Bson, RawBson};
+
+const INT8: u8 = 0x03;
+const FLOAT32: u8 = 0x27;
+const PACKED_BIT: u8 = 0x10;
+
+/// A vector of numeric values. This type can be converted into a [`Binary`] of subtype
+/// [`BinarySubtype::Vector`].
+///
+/// ```rust
+/// # use bson::binary::{Binary, Vector};
+/// let vector = Vector::Int8(vec![0, 1, 2]);
+/// let binary = Binary::from(vector);
+/// ```
+///
+/// `Vector` serializes to and deserializes from a `Binary`.
+///
+/// ```rust
+/// # use serde::{Serialize, Deserialize};
+/// # use bson::{binary::{Result, Vector}, spec::ElementType};
+/// #[derive(Serialize, Deserialize)]
+/// struct Data {
+///     vector: Vector,
+/// }
+///
+/// let data = Data { vector: Vector::Int8(vec![0, 1, 2]) };
+/// let document = bson::to_document(&data).unwrap();
+/// assert_eq!(document.get("vector").unwrap().element_type(), ElementType::Binary);
+///
+/// let data: Data = bson::from_document(document).unwrap();
+/// assert_eq!(data.vector, Vector::Int8(vec![0, 1, 2]));
+/// ```
+///
+/// See the
+/// [specification](https://github.com/mongodb/specifications/blob/master/source/bson-binary-vector/bson-binary-vector.md)
+/// for more details.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Vector {
+    /// A vector of `i8` values.
+    Int8(Vec<i8>),
+
+    /// A vector of `f32` values.
+    Float32(Vec<f32>),
+
+    /// A vector of packed bits. See [`PackedBitVector::new`] for more details.
+    PackedBit(PackedBitVector),
+}
+
+/// A vector of packed bits. This type can be constructed by calling [`PackedBitVector::new`].
+#[derive(Clone, Debug, PartialEq)]
+pub struct PackedBitVector {
+    vector: Vec<u8>,
+    padding: u8,
+}
+
+impl PackedBitVector {
+    /// Construct a new `PackedBitVector`. Each `u8` value in the provided `vector` represents 8
+    /// single-bit elements in little-endian format. For example, the following vector:
+    ///
+    /// ```rust
+    /// # use bson::binary::{Result, PackedBitVector};
+    /// # fn main() -> Result<()> {
+    /// let packed_bits = vec![238, 224];
+    /// let vector = PackedBitVector::new(packed_bits, 0)?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// represents a 16-bit vector containing the following values:
+    ///
+    /// ```text
+    /// [1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0]
+    /// ```
+    ///
+    /// Padding can optionally be specified to ignore a number of least-significant bits in the
+    /// final byte. For example, the vector in the previous example with a padding of 4 would
+    /// represent a 12-bit vector containing the following values:
+    ///
+    /// ```text
+    /// [1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0]
+    /// ```
+    ///
+    /// Padding must be within 0-7 inclusive. Padding must be 0 or unspecified if the provided
+    /// vector is empty.
+    pub fn new(vector: Vec<u8>, padding: impl Into<Option<u8>>) -> Result<Self> {
```

Comment:
the spec requires this type to be valid upon construction, hence the opaque type and result-returning `new` method